// better-auth/dist/adapters/prisma.js
// src/adapters/prisma-adapter/index.ts
function whereConvertor(where) {
  if (!where) return {};
  if (where.length === 1) {
    const w = where[0];
    if (!w) {
      return;
    }
    return {
      [w.field]: w.value
    };
  }
  const and = where.filter((w) => w.connector === "AND" || !w.connector);
  const or = where.filter((w) => w.connector === "OR");
  const andClause = and.map((w) => {
    return {
      [w.field]: w.operator === "eq" || !w.operator ? w.value : {
        [w.operator]: w.value
      }
    };
  });
  const orClause = or.map((w) => {
    return {
      [w.field]: {
        [w.operator || "eq"]: w.value
      }
    };
  });
  return {
    AND: andClause.length ? andClause : void 0,
    OR: orClause.length ? orClause : void 0
  };
}
var prismaAdapter = (prisma, {
  provider
}) => {
  const db = prisma;
  return {
    id: "prisma",
    async create(data) {
      const { model, data: val, select } = data;
      return await db[model].create({
        data: val,
        ...select?.length ? {
          select: select.reduce((prev, cur) => {
            return {
              ...prev,
              [cur]: true
            };
          }, {})
        } : {}
      });
    },
    async findOne(data) {
      const { model, where, select } = data;
      const whereClause = whereConvertor(where);
      return await db[model].findFirst({
        where: whereClause,
        ...select?.length ? {
          select: select.reduce((prev, cur) => {
            return {
              ...prev,
              [cur]: true
            };
          }, {})
        } : {}
      });
    },
    async findMany(data) {
      const { model, where } = data;
      const whereClause = whereConvertor(where);
      return await db[model].findMany({ where: whereClause });
    },
    async update(data) {
      const { model, where, update } = data;
      const whereClause = whereConvertor(where);
      return await db[model].update({
        where: whereClause,
        data: update
      });
    },
    async delete(data) {
      const { model, where } = data;
      const whereClause = whereConvertor(where);
      return await db[model].delete({ where: whereClause });
    }
  };
};
export {
  prismaAdapter
};
