// better-auth/dist/access.js
// src/plugins/organization/access/src/access.ts
var ParsingError = class extends Error {
  path;
  constructor(message, path) {
    super(message);
    this.path = path;
  }
};
var AccessControl = class {
  constructor(s) {
    this.s = s;
    this.statements = s;
  }
  statements;
  newRole(statements) {
    return new Role(statements);
  }
};
var Role = class _Role {
  statements;
  constructor(statements) {
    this.statements = statements;
  }
  authorize(request, connector) {
    for (const [requestedResource, requestedActions] of Object.entries(
      request
    )) {
      const allowedActions = this.statements[requestedResource];
      if (!allowedActions) {
        return {
          success: false,
          error: `You are not allowed to access resource: ${requestedResource}`
        };
      }
      const success = connector === "OR" ? requestedActions.some(
        (requestedAction) => allowedActions.includes(requestedAction)
      ) : requestedActions.every(
        (requestedAction) => allowedActions.includes(requestedAction)
      );
      if (success) {
        return { success };
      }
      return {
        success: false,
        error: `unauthorized to access resource "${requestedResource}"`
      };
    }
    return {
      success: false,
      error: "Not authorized"
    };
  }
  static fromString(s) {
    const statements = JSON.parse(s);
    if (typeof statements !== "object") {
      throw new ParsingError("statements is not an object", ".");
    }
    for (const [resource, actions] of Object.entries(statements)) {
      if (typeof resource !== "string") {
        throw new ParsingError("invalid resource identifier", resource);
      }
      if (!Array.isArray(actions)) {
        throw new ParsingError("actions is not an array", resource);
      }
      for (let i = 0; i < actions.length; i++) {
        if (typeof actions[i] !== "string") {
          throw new ParsingError("action is not a string", `${resource}[${i}]`);
        }
      }
    }
    return new _Role(statements);
  }
  toString() {
    return JSON.stringify(this.statements);
  }
};

// src/plugins/organization/access/statement.ts
var createAccessControl = (statements) => {
  return new AccessControl(statements);
};
var defaultStatements = {
  organization: ["update", "delete"],
  member: ["create", "update", "delete"],
  invitation: ["create", "cancel"]
};
var defaultAc = createAccessControl(defaultStatements);
var adminAc = defaultAc.newRole({
  organization: ["update"],
  invitation: ["create", "cancel"],
  member: ["create", "update", "delete"]
});
var ownerAc = defaultAc.newRole({
  organization: ["update", "delete"],
  member: ["create", "update", "delete"],
  invitation: ["create", "cancel"]
});
var memberAc = defaultAc.newRole({
  organization: [],
  member: [],
  invitation: []
});
var defaultRoles = {
  admin: adminAc,
  owner: ownerAc,
  member: memberAc
};

// src/plugins/organization/access/utils.ts
var permissionFromString = (permission) => {
  return Role.fromString(permission ?? "");
};
export {
  AccessControl,
  ParsingError,
  Role,
  adminAc,
  createAccessControl,
  defaultAc,
  defaultRoles,
  defaultStatements,
  memberAc,
  ownerAc,
  permissionFromString
};
