// better-auth/dist/next-js.js
// src/integrations/next-js.ts
import { betterFetch } from "@better-fetch/fetch";
import { NextResponse } from "next/server";
function toNextJsHandler(auth) {
  const handler = async (request) => {
    return "handler" in auth ? auth.handler(request) : auth(request);
  };
  return {
    GET: handler,
    POST: handler
  };
}
function authMiddleware(options) {
  return async (request) => {
    const url = new URL(request.url).origin;
    const basePath = options?.basePath || "/api/auth";
    const fullURL = `${url}${basePath}/session`;
    const res = await betterFetch(fullURL, {
      headers: {
        cookie: request.headers.get("cookie") || ""
      }
    }).catch((e) => {
      return { data: null };
    });
    const session = res.data || null;
    if (options.customRedirect) {
      return options.customRedirect(session, request);
    }
    if (!session) {
      return NextResponse.redirect(new URL(options.redirectTo || "/", url));
    }
    return NextResponse.next();
  };
}
export {
  authMiddleware,
  toNextJsHandler
};
